### ISSUE TITLE
Segmenting Bacteria in Dense Populations

### ISSUE THREAD

**Issue by ech3000** [lmb-freiburg/Unet-Segmentation#37]
Greetings,
I am trying to apply the U-net algorithm to 2D images of bacteria. The goal is to segment and count bacteria in a dense population. Here is a representative image. This is Acinetobacter Baylyi with a fluorescent DNA stain in green. the pixel width is about 65 nm. The length of such a bacteria is around 1 micron. 
![BF with GFP S4-6](https://user-images.githubusercontent.com/49482753/57139526-c568d200-6db5-11e9-84b2-b3b96a85aeb6.jpg)
Do you think this is feasible? 

To start however, I made low density pictures and annotated them for the U-net to train on. 
![Brightfield with GFP S3-2-C1](https://user-images.githubusercontent.com/49482753/57139968-e3830200-6db6-11e9-81bc-7e95ce010e47.jpg)
When I tell the standard model to segment this kind of images it is not too bad. However, when I finetune with the above shown training data the result of the segmentation after is usually just a black screen. I suspect that my annotations are wrongly interpreted by the algorithm. As info: I have named all 50 ROIs 'cell' and then made an overlay. 
Is my annotation the problem or is it something else? If somebody has an idea how I can improve my training data, e.g. less clumbed up bacteria, different or no staining, different annotations, or bigger bacteria. I would be grateful. 
Here is a list of which settings I have tried: lower learning rate, 100 k and 1 k iterations, annotation via multipoints, zoomed in images, different element sizes, pictures without staining, pictures with only the fluorescent channel, the 'Labels are Classes' box checked and unchecked, and different tile shapes. 
All with the same result a black screen. Below is a figure of what the Loss over Iterations plot of the finetuning usually looks like.
![Loss over Iterations_1000 failed learning](https://user-images.githubusercontent.com/49482753/57141629-ee3f9600-6dba-11e9-9650-b624a1fbbc34.png)

Another issue I have encountered is the re-scaling of the images before segmenting and finetuning. My pictures have a pixel width of about 65 nm. I have changed the properties of the image accordingly. When I segment the first thing the algorithm does is re-scale it to 0.5 by 0.5 microns. As a result it blurs the image to an almost unusable point. Below is the re-scaled version of the second image. 
![Brightfield with GFP S3-2-C1 - 32-Bit - rescaled (xy) - normalized](https://user-images.githubusercontent.com/49482753/57144283-087c7280-6dc1-11e9-8178-ffc4659af04f.jpg)
I can work around this by setting the pixel width to 0.5 by 0.5 microns. However while finetuning, I will have to set the pixel width and the element size both to 0.5 by 0.5. At this point I am not sure if the algorithm then looks for elements which are one pixel in width. Does it do that? What is this re-scaling about and good for?

I am thankful for any answer to one of my many questions. 

Best, 
Reto

PS @ Thorsten: I managed to install U-net on a different computer running ubuntu 19.04.

**Comment by ThorstenFalk** [comment]
Dear Reto,

In general I would say it is feasible, but the density and size of your bacteria makes it a hard task to properly separate them all. The 2d_cell_net_v0 model was trained on images with 0.5µm/px, which is of course not suitable for your images. Your workaround to trick the plugin should lead to some non-empty segmentation result out-of-the-box, but the model will not be able to separate individual bacteria. Finetuning on subsets of the high-density images is required, so that the network sees sufficiently many samples of densely packed bacteria. I would recommend to select one or two rectangular regions in at least two images, and turning the rest into an ignore-ROI. Annotate all bacteria in the rectangles as you did in the low-density case. Keep the original element size of the images and click "From image" in the element size selection, this will use the raw image element size (of the first image in the training list) for the new model. When applying the finetuned model to new images they will then be rescaled to that element size instead. This rescaling is implemented basically to exploit the known actual scales of the objects to segment.

With these steps I'd assume at least reasonable results.

All the best,

Thorsten

**Comment by ech3000** [comment]
Hey Thorsten,
I have achieved a very good segmentation of the first image. Few things did it for me: Creating a new model and training from scratch, annotating via mask as described in the suppl. info of your paper, and as you suggested keeping the original pixel sizes and clicking "From Image". 
Surprisingly, a training data set with high density bacteria was not needed to achieve a very good segmentation. But I will keep it in mind for when I want an even better segmentation. Thank you very much for your reply and your advice. 
Best,
Reto

**Comment by ech3000** [comment]
![BF S4-6 tif - segmented ADP1_100000IT_mask_075scale_data6x2](https://user-images.githubusercontent.com/49482753/58251863-0de82f80-7d65-11e9-8cb1-a95e580e130a.jpg)

This is what my result looks like. In case someone is curious. I trained for 100'000 iterations on 8 training images like the one shown above. Except that I shrunk the ROI by (0.75)^2 in order to get a distinct segmentation.

**Comment by ThorstenFalk** [comment]
Shrinking the ROIs is actually a good trick if you are basically interested in a detection/counting task! Some bacteria chains are not properly segmented, but on a glance a very nice result!
